document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -123,13 +125,66 @@ public boolean upsert(Key key, Document document) throws IOException {
         LOGGER.debug("Write result: " + writeResult.toString());
       }
 
-      return writeResult.getModifiedCount() > 0;
+      return (writeResult.getUpsertedId() != null || writeResult.getModifiedCount() > 0);
     } catch (IOException e) {
       LOGGER.error("Exception upserting document. key: {} content:{}", key, document, e);
       throw e;
     }
   }
 
+  /**
+   * Update an existing document if condition is evaluated to true. Conditional will help in
+   * providing optimistic locking support for concurrency update.
+   */
+  @Override
+  public DocStoreResult update(Key key, Document document, Filter condition) throws IOException {
+    try {
+      DocStoreResult.Builder resultBuilder = DocStoreResult.Builder.newBuilder();
+      Map<String, Object> conditionMap =
+          condition != null ? parseQuery(condition) : new HashMap<>();
+      conditionMap.put(ID_KEY, key.toString());
+      BasicDBObject conditionObject = new BasicDBObject(conditionMap);
+      UpdateOptions options = new UpdateOptions().upsert(false);
+
+      UpdateResult writeResult =
+          collection.updateOne(conditionObject, this.prepareUpsert(key, document), options);
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Update result: " + writeResult.toString());
+      }
+      boolean success = writeResult.getModifiedCount() > 0;
+      resultBuilder.setSuccess(success).setResultCount(writeResult.getModifiedCount());
+      if (!success) {
+        resultBuilder.setFailureMessage("Failed to update document with key:" + key.toString());
+      }
+      return resultBuilder.build();
+    } catch (Exception e) {
+      LOGGER.error("Exception updating document. key: {} content:{}", key, document, e);
+      throw new IOException(e);
+    }
+  }
+
+  /** create a new document if one doesn't exists with key */
+  @Override
+  public DocStoreResult create(Key key, Document document) throws IOException {
+    try {
+      DocStoreResult.Builder resultBuilder = DocStoreResult.Builder.newBuilder();
+      BasicDBObject basicDBObject = this.prepareInsert(key, document);
+      InsertOneResult insertOneResult = collection.insertOne(basicDBObject);
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Create result: " + insertOneResult.toString());
+      }
+      boolean success = insertOneResult.getInsertedId() != null;
+      resultBuilder.setSuccess(success).setResultCount(1);
+      if (!success) {
+        resultBuilder.setFailureMessage("Failed to create document with key:" + key.toString());
```

Comment:
should the original error message be captured and included here ? for example - duplicate error or something else ? The doc for insertOne actually throws an exception - which I'm assuming InsertOneResult is capturing ?

Reply:
If there is an exception, that will be wrapped into IOException and thrown to the client. The duplicate error is an exception scenario. Here, we are handling the case where there was no exception.